machine-learning/01-openpbta_heatmap.Rmd

```diff
@@ -0,0 +1,383 @@
+# Row indices of high variance genes
+high_variance_index <- which(gene_variance > variance_threshold)
+# What does a row index look like?
+head(high_variance_index)
+```
+
+```{r high_var_mat, live = TRUE}
+# Get a matrix that is subset to just the high variance genes
+high_var_mat <- rnaseq_mat[high_variance_index, ]
+```
+
+## Heatmap
+
+### Annotation
+
+First, we'll set up the metadata that we want to use to label samples in the heatmap.
+In `ComplexHeatmap` terminology, this is called annotation, or `HeatmapAnnotation`, specifically.
+
+```{r sample_annotation_df}
+sample_annotation_df <- filtered_histologies_df %>%
+  # Select only the columns that we'll use
+  select(Kids_First_Biospecimen_ID,
+         short_histology,
+         composition,
+         tumor_descriptor)
+
+# Let's examine these columns
+sample_annotation_df
+```
+
+Right now, `sample_annotation_df` is a tibble because it's a filtered version of something we read in with `readr::read_tsv()`.
+Usually this is pretty much equivalent to a data frame.
+
+```{r sample_annotation_class}
+class(sample_annotation_df)
+```
+
+BUT `ComplexHeatmap` expects a data frame, so sometimes it will behave in unexpected ways when you give it a tibble! 
+Let's make a data frame and make the sample identifiers the rownames.
+
+```{r annotation_as_df, live = TRUE}
+sample_annotation_df <- sample_annotation_df %>%
+  as.data.frame() %>%
+  tibble::column_to_rownames("Kids_First_Biospecimen_ID")
+```
+
+To specify the colors in our annotation bar, we need to create a list of named vectors.
+The names of the list need to exactly match the column names in `sample_annotation_df` and the names in each vector need to exact match the values in those columns.
+
+```{r color_palettes}
+# Palette recommended for those with color vision deficencies - we'll use that 
+# for histologies
+histology_colors <- colorblindr::palette_OkabeIto[2:5]
+sample_annotation_colors <- list(
+  short_histology = c(
+    "LGAT" = histology_colors[1],
+    "Ependymoma" = histology_colors[2],
+    "HGAT" = histology_colors[3],
+    "Medulloblastoma" = histology_colors[4]
+  ),
+  composition = c(
+    "Solid Tissue" = "#A0A0A0",  # light grey
+    "Derived Cell Line" = "#000000"  # black
+  ),
+  tumor_descriptor = c(
+    "Initial CNS Tumor" = "#3333FF",
+    "Progressive" = "#FFFF99",
+    "Recurrence" = "#CCCCFF",
+    "Second Malignancy" = "#000033",
+    "Unavailable" = "#FFFFFF"  # white for missing data
+  )
+)
+```
+
+We need to create a special type of object with `HeatmapAnnotation`.
+
+```{r create_annotation, live = TRUE}
+column_annotation <- HeatmapAnnotation(
+  df = sample_annotation_df,
+  col = sample_annotation_colors
+)
+```
+
+### Values for display
+
+We will z-score the expression values for display.
+This is sometimes called a standardized score.
+
+```{r row_normalize}
+# PLIER::rowNorm z-scores the gene (row) values
+zscores_mat <- PLIER::rowNorm(high_var_mat)
```

Comment:
If we want to include an alternative that follows the Z score formula:
```
(high_var_mat - rowMeans(high_var_mat)) / matrixStats::rowSds(high_var_mat)
```

(don't try this with column z scores!)

Reply:
Oh I thought for sure the `sweep()`-based solution would be faster, but that is not the case (my guess is `matrixStats::rowSds()` and `rowMeans()` is why this is faster).